### PR TITLE
zHash fork

### PR DESCRIPTION
**Pull request by davidTjr**
Review needed for changes.

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -61,10 +61,11 @@ class CMainParams : public CChainParams {
         //nMinerThreads = 0;
         nMaxTipAge = 24 * 60 * 60;
         nPruneAfterHeight = 100000;
-        const size_t N = 200, K = 9;
-        BOOST_STATIC_ASSERT(equihash_parameters_acceptable(N, K));
-        nEquihashN = N;
-        nEquihashK = K;
+        eh_epoch_1 = eh200_9;
+        eh_epoch_2 = eh144_5;
+        eh_epoch_1_endblock = 95010;
```
Comment (renuzit) [inline src/chainparams.cpp]: To be honest, I wasn't a fan of having an overlap period. If you want you could set it to 1 block or something.

File: src/main.cpp [hunk 2]
```diff
@@ -3328,7 +3328,28 @@ bool ContextualCheckBlockHeader(const CBlockHeader& block, CValidationState& sta
 
     assert(pindexPrev);
 
-    int nHeight = pindexPrev->nHeight+1;
+    long int nHeight = pindexPrev->nHeight+1;
```
Comment (renuzit) [inline src/main.cpp]: Should probably be:
`int nHeight = pindexPrev->nHeight+1;` and not a `long int`

File: src/crypto/equihash.cpp [hunk 3]
```diff
@@ -36,7 +36,10 @@ int Equihash<N,K>::InitialiseState(eh_HashState& base_state)
     uint32_t le_N = htole32(N);
     uint32_t le_K = htole32(K);
     unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES] = {};
-    memcpy(personalization, "ZcashPoW", 8);
+    if(N==144 && K==5)
+        memcpy(personalization, "Vision", 8);
```
Comment (renuzit) [inline src/crypto/equihash.cpp]: If you want this compatible with the ZHash miner, it will need to be set to `BitcoinZ`

Comment (renuzit) [reply src/crypto/equihash.cpp]: If not, you'll need to add 2 more chars. This must be 8 characters.

File: src/sendalert.cpp [hunk 4]
```diff
@@ -82,7 +82,7 @@ void ThreadSendAlert()
     // These versions are protocol versions
     // 170002 : 1.0.0
     alert.nMinVer       = 170002;
```
Comment (renuzit) [inline src/sendalert.cpp]: For public release `alert.nMinVer` should be set to 170004, however, you will need to setup a few nodes that you control with the nMinVer set to 170002 to act as a bridge until the network upgrade has started, then they need to be switched over to the client release.

File: configure.ac [hunk 5]
```diff
@@ -1,7 +1,7 @@
 dnl require autoconf 2.60 (AS_ECHO/AS_ECHO_N)
 AC_PREREQ([2.60])
 define(_CLIENT_VERSION_MAJOR, 1)
-define(_CLIENT_VERSION_MINOR, 0)
+define(_CLIENT_VERSION_MINOR, 1)
 define(_CLIENT_VERSION_REVISION, 14)
```
Comment (renuzit) [inline configure.ac]: CLIENT_VERSION_REVISION is probably 0, since this is a new Minor release. ;)

File: src/clientversion.h [hunk 6]
```diff
@@ -17,7 +17,7 @@
 
 //! These need to be macros, as clientversion.cpp's and bitcoin*-res.rc's voodoo requires it
 #define CLIENT_VERSION_MAJOR 1
-#define CLIENT_VERSION_MINOR 0
+#define CLIENT_VERSION_MINOR 1
 #define CLIENT_VERSION_REVISION 14
```
Comment (renuzit) [inline src/clientversion.h]: CLIENT_VERSION_REVISION is probably 0, since this is a new Minor release. ;)

**Comment by renuzit** [comment]
I just had a few nit picks. Fix these or respond to the feedback and let me know when you're ready to look at this again.

File: configure.ac (hunk 5 above)
Comment (davidTjr) [reply configure.ac]: 💯 true, this is just easier for me to keep track ;)

File: src/clientversion.h (hunk 6 above)
Comment (davidTjr) [reply src/clientversion.h]: yup, as above.

**Comment by davidTjr** [comment]
@renuzit changes made, is there anything else to watch out for?

File: src/pow.cpp [hunk 7]
```diff
@@ -22,12 +22,20 @@
...
 unsigned int GetNextWorkRequired(const CBlockIndex* pindexLast, const CBlockHeader *pblock, const Consensus::Params& params)
 {
+    const CChainParams& chainParams = Params();
     unsigned int nProofOfWorkLimit = UintToArith256(params.powLimit).GetCompact();
 
     // Genesis block
     if (pindexLast == NULL)
         return nProofOfWorkLimit;
 
+    // Reset the difficulty after the algo fork
```
Comment (renuzit) [inline src/pow.cpp]: Remove the difficulty reset. EBWF is efficient enough for this not to matter anymore, unless you want 13 really fast blocks. I could also link you to what BTG did, it's a bit better of a job resetting the difficulty. They took the currently and divided it by 100.

File: src/sendalert.cpp [hunk 8]
```diff
@@ -78,11 +78,12 @@ void ThreadSendAlert()
     alert.nID           = 1003;  // use https://github.com/MyVision/vision/wiki/Alerts to keep track of ids
     alert.nCancel       = 1001;  // cancels previous messages up to this ID number
 
-    // Currently we are exactly the same as our Upstream Zcash protocol version
     // These versions are protocol versions
     // 170002 : 1.0.0
+    // 170004 : 1.1.0
+
     alert.nMinVer       = 170002;
```
Comment (renuzit) [inline src/sendalert.cpp]: I would set MinVer to 170004.

File: src/pow.cpp (hunk 7 above)
Comment (renuzit) [reply src/pow.cpp]: Never-mind, I think we've come to a consensus that resetting the difficulty is the best thing to do here.

Comment (davidTjr) [reply src/pow.cpp]: ok. thx 👍

**Pull request merged**